hypertrace-graphql-labels-schema/src/main/java/org/hypertrace/graphql/label/joiner/LabelJoiner.java

```diff
@@ -0,0 +1,43 @@
+package org.hypertrace.graphql.label.joiner;
+
+import io.reactivex.rxjava3.core.Single;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.hypertrace.graphql.label.schema.LabelResultSet;
+
+public interface LabelJoiner {
+
+  /** A NOOP joiner */
+  LabelJoiner NO_OP_JOINER =
+      new LabelJoiner() {
+        @Override
+        public <T> Single<Map<String, LabelResultSet>> joinLabels(
+            Collection<T> joinSources, IdGetter<T> idGetter, LabelsGetter<T> labelsGetter) {
+          return Single.just(Collections.emptyMap());
+        }
+      };
+
+  /**
+   * Produces a map of label result set to source ids
+   *
+   * @param joinSources a collection of source data
+   * @param idGetter A method that retrieves an ID from the source
+   * @param labelsGetter A method that retrieves labels from the source
+   * @param <T> Type of source data
+   * @return A map of id of each source to its matching label result set
+   */
+  <T> Single<Map<String, LabelResultSet>> joinLabels(
+      Collection<T> joinSources, IdGetter<T> idGetter, LabelsGetter<T> labelsGetter);
+
+  @FunctionalInterface
+  interface IdGetter<T> {
```

Comment:
I'm a little confused on how this is supposed to work. Why do we need an id from the source? are we using that in lieu of the original object on the returned map (if so, why)? And is getLabels getting us the label IDs off the entities? It seems like that's all we really need. Basically just:

```java
public interface LabelJoiner {
  <T> Single<Map<T, LabelResultSet>> joinLabels(
      Collection<T> joinSources, LabelIdGetter<T> idGetter);

  @FunctionalInterface
  interface LabelIdGetter<T> {
    Single<List<String>> getLabelIds(T source); // or sync, async just gives us more flexibility in impl
  }
}
```

Reply:
The assumption you are making is that source implements `equals` and `hashCode` methods. The id getter can just return the entity itself.

Anyways since it will be used to just join labels and entities, I can get rid of IdGetter.